### PR TITLE
8257241: CDS should not handle disableEagerInitialization for archived lambda proxy classes

### PR DESCRIPTION
**Pull request by calvinccheung**
Please review this change which includes:

- If the `jdk.internal.lambda.disableEagerInitialization`property is enabled, the `InnerClassLambdaMetafactory` will not involve CDS to archive lambda proxy classes or to find them from an archive.

- Not passing the `initialize` (same as `!disableEagerInitialization`) to `LambdaProxyClassArchive.find` and eventually to the `JVM_LookupLambdaProxyClassFromArchive` function.

Testing:

- [x] all cds/appcds tests locally on linux-x64

- [x] tiers 1 - 4 (in progress)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257241](https://bugs.openjdk.java.net/browse/JDK-8257241): CDS should not handle disableEagerInitialization for archived lambda proxy classes

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to fc10e77fd03ffce6c7efdb5bcbed7922c73f14d2
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to 812777b3e69bfceea328df34149cb0964aa8625a
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to fc10e77fd03ffce6c7efdb5bcbed7922c73f14d2

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1542/head:pull/1542`
`$ git checkout pull/1542`

### CODE REVIEW

**Comment by mlchung** [comment]
The change looks okay.   Can you add a test to verify that it won't load from CDS archive if `disableEagerInitialization` is true?

**Review by iklam — COMMENTED** [review]
LGTM. I agree with Mandy that we need a test case.

**Review by cl4es — APPROVED** [review]
Looks good to me!

**Comment by calvinccheung** [comment]
@mlchung, @iklam, @cl4es Thanks for the review.
I've added a test.

File: test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * @bug 8257241
+ * @summary Run the LambdaEagerInitTest.java test in static CDS archive mode.
+ *          Create a base archive with the -Djdk.internal.lambda.disableEagerInitialization=true property.
+ *          Run with the archive with and without specifying the property.
+ *          With the disableEagerInit set to true, lambda proxy classes will not
+ *          be archived. During runtime, lambda proxy classes will not be loaded
+ *          from the archive.
+ * @requires vm.cds
+ * @library /test/lib /test/hotspot/jtreg/runtime/cds/appcds
+ * @compile ../../../../../jdk/java/lang/invoke/lambda/LambdaEagerInitTest.java
```
Comment (mlchung) [inline test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java:36]: I think it's better to make a copy of this test along side with this test.

File: test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            .addPrefix("-cp", appJar, testProperty,  "-Xlog:class+load,cds=debug")
+            .setArchiveName(archiveName)
+            .setUseVersion(false)
+            .addSuffix(mainClass);
+        OutputAnalyzer output = CDSTestUtils.runWithArchive(runOpts);
+        output.shouldMatch(".class.load. java.util.stream.Collectors[$][$]Lambda[$].*/0x.*source:.*java.*util.*stream.*Collectors")
+              .shouldNotMatch(".cds.*Loaded.*lambda.*proxy")
+              .shouldHaveExitValue(0);
+
+        // run with archive without the -Djdk.internal.lambda.disableEagerInitialization=true property
```
Comment (mlchung) [inline test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java:77]: Minor suggestion: make this test case in a separate `testWithEagerInitializationEnabled` method.   `testImpl` can be renamed to `testWithEagerInitializationDisabled`.

File: test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java (hunk 1 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java:36]: I can make a copy and put it under the `open/test/hotspot/jtreg/runtime/cds/appcds/test-classes` dir. I'd still need the `@compile` and would be something like `@compile test-classes/LambdaEagerInitTest.java`.

Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java:36]: The `@compile` can be avoiding by putting LambdaEagerInitTest into the `@library` and referencing it as

```
private static final String mainClass = LambdaEagerInitTest.class.getName();
```

Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java]: Yes, `@library test-classes` works.

File: test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java (hunk 2 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/LambdaEagerInit.java:86]: I also factored out the archive creation into the `createArchiveWithEagerInitializationEnabled` method.

File: test/hotspot/jtreg/runtime/cds/appcds/test-classes/LambdaEagerInitTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
```
Comment (mlchung) [inline test/hotspot/jtreg/runtime/cds/appcds/test-classes/LambdaEagerInitTest.java:25]: This is part of LambdaEagerInit test.   This does not need to be `@test` and so the entire comment block can be removed.

If you want to verify, you could add a test case in LambdaEagerInit.java to run this main class without the CDS archive.

**Review by iklam — APPROVED** [review]

**Review by mlchung — APPROVED** [review]

**Pull request closed without merging**